src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -516,7 +516,7 @@ private void OnBatchScopeDisposed()
                         {
                             // TODO: find a cleaner way to fetch this
                             var metadataReference = _workspace.CurrentSolution.GetRequiredProject(Id).MetadataReferences.Cast<PortableExecutableReference>()
-                                                                                    .Single(m => m.FilePath == path && m.Properties == properties);
+                                .Single(m => StringComparer.OrdinalIgnoreCase.Equals(m.FilePath, path) && m.Properties == properties);
```

Comment:
We have unit tests for this, so if this is indeed the bug we should easily be able to add tests under https://github.com/dotnet/roslyn/tree/main/src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests.

However, the fix might require more investigation: I see in https://github.com/dotnet/roslyn/blob/d695369176aa109f0e0fb8b0019294aaa0823a54/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs#L1868 that we are also doing a path-sensitive check there. I'm also suspicious generally of case-insenstivity like this -- we've had requests from VS for Mac to share this code cross-plat, and it'd have to be case sensitive there.

Reply:
i don't like this as this seems to be masking a bug elsewhere.  what is the scenario where we get teh string with one casing in one path, but then it comes through with another casing in another?

Reply:
It also shouldn't have gotten that far either:

https://github.com/dotnet/roslyn/blob/d695369176aa109f0e0fb8b0019294aaa0823a54/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs#L965-L968

That is doing a case sensitive check.

Reply:
Ah, I bet the problem is actually here:

https://github.com/dotnet/roslyn/blob/d695369176aa109f0e0fb8b0019294aaa0823a54/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs#L965-L968

For purposes of converting metadata references to project references, we are letting that be case insensitive. Ignoring for a moment whether that's masking other bugs, I wonder if there's a bug here where when we convert a project reference back (because the target project is removed, for example), that we don't restore it with the original case. Trying that now.

Reply:
Alright past me was smarter than current me.* It looks like we do correctly restore the original path casing in that scenario.

I've sent out another PR (https://github.com/dotnet/roslyn/pull/56490) and added more tests here, but those seem to pass. I'm stumped again.

(*I am reminded of the quote that "debugging code is twice as hard as writing it; if you write code as cleverly as possible, you aren't smart enough to debug it"...although in this case I had hoped the code was as dumb as possible!)